compiler_opt/es/blackbox_evaluator.py

```diff
@@ -101,5 +101,62 @@ def get_results(
 
     return futures
 
-  def set_baseline(self) -> None:
+  def set_baseline(self, pool: FixedWorkerPool) -> None:
+    del pool  # Unused.
     pass
+
+
+@gin.configurable
+class TraceBlackboxEvaluator(BlackboxEvaluator):
+  """A blackbox evaluator that utilizes trace based cost modelling."""
+
+  def __init__(self, train_corpus: corpus.Corpus,
+               est_type: blackbox_optimizers.EstimatorType, bb_trace_path: str,
+               function_index_path: str):
+    self._train_corpus = train_corpus
+    self._est_type = est_type
+    self._bb_trace_path = bb_trace_path
+    self._function_index_path = function_index_path
+
+    self._has_baseline = False
+    self._baseline = 1
+
+  def get_results(
+      self, pool: FixedWorkerPool, perturbations: List[policy_saver.Policy]
+  ) -> List[concurrent.futures.Future]:
+    job_args = []
+    for perturbation in perturbations:
+      job_args.append(
+          (self._train_corpus.module_specs, self._function_index_path,
+           self._bb_trace_path, perturbation))
+
+    _, futures = buffered_scheduler.schedule_on_worker_pool(
+        action=lambda w, v: w.compile_corpus_and_evaluate(
+            v[0], v[1], v[2], v[3]),
+        jobs=job_args,
+        worker_pool=pool)
+    concurrent.futures.wait(
+        futures, return_when=concurrent.futures.ALL_COMPLETED)
+    return futures
+
+  def set_baseline(self, pool: FixedWorkerPool) -> None:
+    if self._has_baseline:
```

Comment:
can we make re-setting the baseline throw, wouldn't it be a bug? the user can just check if it was set

alternatively, could it be set at ctoring time? probably not, but wanted to check

Reply:
I had it set up originally where it just gets called every iteration. There's no real reason to do that though, so I've updated it to throw an error if it is called while the baseline is already set.

We need access to the pool to dispatch a job to get the baseline, and we don't have access in the pool in the constructor, so we cannot.